timemachine/cpp/src/local_md_potentials.hpp

```diff
@@ -0,0 +1,66 @@
+#pragma once
+#include "device_buffer.hpp"
+#include "pinned_host_buffer.hpp"
+#include <memory>
+#include <vector>
+
+#include "bound_potential.hpp"
+#include "curand.h"
+#include "flat_bottom_bond.hpp"
+#include "local_md_utils.hpp"
+
+namespace timemachine {
+
+class LocalMDPotentials {
+
+public:
+    LocalMDPotentials(const int N, const std::vector<std::shared_ptr<BoundPotential>> bps);
+
+    ~LocalMDPotentials();
+
+    DeviceBuffer<unsigned int> *get_free_idxs();
+
+    void setup_from_idxs(
+        double *d_x_t_,
+        double *d_box_t,
+        const std::vector<int> &local_idxs,
+        const double temperature,
+        const int seed,
+        const double radius,
+        const double k,
+        const cudaStream_t stream);
+
+    std::vector<std::shared_ptr<BoundPotential>> get_potentials();
+
+    void reset(const cudaStream_t stream);
+
+private:
+    const int N_;
+    std::size_t temp_storage_bytes_;
+
+    std::vector<std::shared_ptr<BoundPotential>> all_potentials_;
+    std::shared_ptr<BoundPotential> ixn_group_;
+    std::shared_ptr<BoundPotential> nonbonded_bp_;
+    std::shared_ptr<FlatBottomBond<double>> restraint_;
+    std::shared_ptr<BoundPotential> bound_restraint_;
+
+    DeviceBuffer<int> restraints_;
```

Comment:
Rename to `restraint_idxs_` or `restraint_pairs_`?

Reply:
(also currently very easy to typo into the shared_ptr `restraint_`)

Reply:
Good catch, fixed here: https://github.com/proteneer/timemachine/pull/1004/commits/59ead18e83f111b194321725b0d92782388c468b